src/player/index.ts

```diff
@@ -51,628 +192,567 @@ export class Player extends GObject.Object {
-      cb();
-    }
-  }
+    // volume
 
-  private async_done_cbs: (() => void)[] = [];
+    Settings.connect("changed::volume", () => {
+      const settings_volume = Settings.get_double("volume");
+      if (settings_volume !== this.volume) {
+        this.volume = settings_volume;
+      }
+    });
 
-  private add_async_done_cb(cb: () => void) {
-    this.async_done_cbs.push(cb);
+    this.volume = Settings.get_double("volume");
+
+    this.connect("notify::volume", () => {
+      const settings_volume = Settings.get_double("volume");
+      if (settings_volume !== this.volume) {
+        Settings.set_double("volume", this.volume);
+      }
+    });
+
+    // restore state
+
+    this.load_state();
   }
 
-  private fire_async_done_cbs() {
-    const cbs = this.async_done_cbs;
-    this.async_done_cbs = [];
+  // property: current-meta
 
-    for (const cb of cbs) {
-      cb();
-    }
+  private _now_playing: ObjectContainer<TrackMetadata> | null = null;
+
+  get now_playing() {
+    return this._now_playing;
   }
 
-  private on_message(_bus: Gst.Bus, message: Gst.Message) {
-    const type = message.type;
+  // whether the next play action should add to the history
+  private add_history = false;
 
-    const [position_available, position] = this.playbin.query_position(
-      Gst.Format.TIME,
-    );
+  play() {
+    super.play();
 
-    if (position_available) {
-      this.position = position;
-    }
+    // TODO: add history
+    // const id = this.now_playing?.object.track.videoId;
 
-    switch (type) {
-      case Gst.MessageType.EOS:
-        this.repeat_or_next();
-        break;
-      case Gst.MessageType.ERROR:
-        this.stop();
-        const [error, debug] = message.parse_error();
-        console.log("Error: ", error, debug);
-      case Gst.MessageType.CLOCK_LOST:
-        this.playbin.set_state(Gst.State.PAUSED);
-        this.playbin.set_state(Gst.State.PLAYING);
-        break;
-      case Gst.MessageType.BUFFERING:
-        const percent = message.parse_buffering();
+    // if (id) {
+    //   if (this.add_history && get_option("auth").has_token()) {
+    //     // add history entry, but don't wait for the promise to resolve
+    //     add_history_item(id);
 
-        if (this._is_live) return;
+    //     this.add_history = false;
+    //   }
+    // }
+  }
 
-        if (percent < 100) {
-          if (!this.buffering && this.playing) {
-            this.playbin.set_state(Gst.State.PAUSED);
-          }
+  private loading_controller: AbortController | null = null;
 
-          this.buffering = true;
-        } else {
-          this.buffering = false;
+  private _loading_track: string | null = null;
 
-          if (this.playing) this.playbin.set_state(Gst.State.PLAYING);
-        }
-        break;
-      case Gst.MessageType.ASYNC_DONE:
-        this.fire_async_done_cbs();
-        break;
-    }
+  get loading_track() {
+    return this._loading_track ?? null;
```

Comment:
```suggestion
    return this._loading_track;
```

At no point should it be `undefined`?